ext/oj/parse.c

```diff
@@ -327,26 +327,82 @@ static void read_escaped_str(ParseInfo pi, const char *start) {
     buf_cleanup(&buf);
 }
 
-static void read_str(ParseInfo pi) {
-    const char *str    = pi->cur;
-    Val         parent = stack_peek(&pi->stack);
-
+static int scan_string_noSIMD(ParseInfo pi, const char *str) {
     for (; '"' != *pi->cur; pi->cur++) {
         if (pi->end <= pi->cur) {
             oj_set_error_at(pi,
                             oj_parse_error_class,
                             __FILE__,
                             __LINE__,
                             "quoted string not terminated");
-            return;
+            return false;
         } else if ('\0' == *pi->cur) {
             oj_set_error_at(pi, oj_parse_error_class, __FILE__, __LINE__, "NULL byte in string");
-            return;
+            return false;
         } else if ('\\' == *pi->cur) {
             read_escaped_str(pi, str);
-            return;
+            return false;
+        }
+    }
+
+    return true;
+}
+
+#if defined(OJ_USE_SSE4_2)
+#include <nmmintrin.h>
+
+static int scan_string_SIMD(ParseInfo pi, const char *str) {
+    static const char chars[16] = "\"\\\0";
+    const __m128i terminate = _mm_loadu_si128((const __m128i *)&chars[0]);
+    const char *end = (const char *)(pi->end - 16);
+
+    for (; pi->cur <= end; pi->cur += 16) {
+        const __m128i string = _mm_loadu_si128((const __m128i *)pi->cur);
+        const int r = _mm_cmpistri(terminate, string, _SIDD_UBYTE_OPS | _SIDD_LEAST_SIGNIFICANT | _SIDD_CMP_RANGES);
+        if (r != 16) {
+            pi->cur = pi->cur + r;
+
+            if (pi->end <= pi->cur) {
+                oj_set_error_at(pi,
+                                oj_parse_error_class,
+                                __FILE__,
+                                __LINE__,
+                                "quoted string not terminated");
+                return false;
+            } else if ('"' == *pi->cur) {
```

Comment:
Just make this:
```}
if ('"' == *pi->cur) {
```
Same for the next few blocks. If there is a return there is no reason for an else